src/witness_generator/generator.rs

```diff
@@ -263,10 +263,12 @@ where
         } else {
             evaluated
                 .solve_with_bit_constraints(&self.bit_constraint_set())
-                .map_err(|_| {
+                .map_err(|e| {
                     let formatted = evaluated.format(self.fixed_data);
-                    if evaluated.is_invalid() {
-                        format!("Constraint is invalid ({formatted} != 0).").into()
+                    if let EvalError::ConstraintUnsatisfiable(_) = e {
+                        EvalError::ConstraintUnsatisfiable(format!(
+                            "Constraint is invalid ({formatted} != 0)."
```

Comment:
It seems to me that this message can be put in `solve` already. We have access to the expression there as well and it would avoid this rewrite.

Reply:
Oh is it because the solver does not really deal with "constraints"?

Reply:
The affine expression does not know the names of the variables. It needs access to a WitnessColumnNamer like in line 267 to properly format.